File: Leetcode/Problems C++/105. Construct Binary Tree from Preorder and Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTreeHelp(vector<int>& preorder, vector<int>& inorder,int psi,int pei,int isi,int iei)
    {
        if(psi>pei || isi>iei)
            return nullptr;
        TreeNode* root = new TreeNode(preorder[psi]);
        int rootIndex=0;
        
        while(inorder[rootIndex]!=preorder[psi]) rootIndex++;
        
        int lpsi = psi+1;
        int lisi = isi;
        int liei = rootIndex-1;
        int lpei = liei - lisi + lpsi;
        
        int rpsi = lpei +1;
        int rpei = pei;
        int risi = rootIndex+1;
        int riei = iei;
        
        root->left = buildTreeHelp(preorder,inorder,lpsi,lpei,lisi,liei);
        root->right = buildTreeHelp(preorder,inorder,rpsi,rpei,risi,riei);
        return root;
        
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return buildTreeHelp(preorder,inorder,0,preorder.size()-1,0,inorder.size()-1);
    }
};

File: Leetcode/Problems Java/105. Construct Binary Tree from Preorder and Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder,int sp, int ep, int[] inorder, int si, int ei){
        if(sp>ep || si>ei)return null;
        
        TreeNode root = new TreeNode(preorder[sp]);

        int index = si;
        while(preorder[sp] != inorder[index]){
            index+=1;
        }
        
        int leftStart_ip = si;
        int leftEnd_ip = index-1;
        int leftStart_sp =  sp+1;
        int leftEnd_sp = leftEnd_ip - leftStart_ip + leftStart_sp;
        
        int rightStart_ip = index+1;
        int rightEnd_ip = ei;
        int rightStart_sp =  leftEnd_sp+1;
        int rightEnd_sp = ep;
        
        root.left = buildTree(preorder, leftStart_sp, leftEnd_sp, inorder, leftStart_ip, leftEnd_ip);
        root.right = buildTree(preorder, rightStart_sp, rightEnd_sp, inorder, rightStart_ip, rightEnd_ip);
        return root;
    }
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return buildTree(preorder, 0, preorder.length-1, inorder, 0, inorder.length-1);
    }
}
